### DISCUSSION TITLE
Storage RLS Performance: Index not used with RLS enabled

### DISCUSSION THREAD

**Discussion by psteinroe** [supabase/supabase#9207] [Questions]
Hey everyone,

I am having major struggles with the query performance of the storage.objects table and cannot get my head around it. Any help is highly appreciated.

We are providing a multi-tenant solution where different entities can have files attached. To easily query them on different levels, we enforce the following directory structure:
`organisation-id/entity-id/myfile.jpg`

In some cases, we also have one more level, e.g. for conversation with messages, we want to attach files to a message but also query all files of a conversation:  
`organisation-id/conversation-id/message-id/myfile.jpg`.

To query files for an entity, we use the following query (simplified):
```sql
select id
from storage.objects
where bucket_id = 'message_files'
and uuid_from_path_level_2(name) = '7daa4a53-1233-432d-bf11-a9dee8140027'::uuid;
``` 

For a better performance, we setup a bunch of indices, and when executed as service_role (bypassing rls), the indices are properly picked up ([Visualised on explain.dalibo](https://explain.dalibo.com/plan/4247g6e82726c139)).
```json
{
    "Plan": {
      "Node Type": "Index Scan",
      "Parallel Aware": false,
      "Async Capable": false,
      "Scan Direction": "Forward",
      "Index Name": "ix_storage_obj_entity_id_from_path",
      "Relation Name": "objects",
      "Schema": "storage",
      "Alias": "objects",
      "Startup Cost": 0.42,
      "Total Cost": 370.44,
      "Plan Rows": 706,
      "Plan Width": 16,
      "Actual Startup Time": 0.032,
      "Actual Total Time": 0.036,
      "Actual Rows": 2,
      "Actual Loops": 1,
      "Output": ["id"],
      "Index Cond": "((objects.bucket_id = 'message_files'::text) AND (uuid_from_path_level_2(objects.name) = '7daa4a53-1233-432d-bf11-a9dee8140027'::uuid))",
      "Rows Removed by Index Recheck": 0,
      "Shared Hit Blocks": 5,
      "Shared Read Blocks": 0,
      "Shared Dirtied Blocks": 0,
      "Shared Written Blocks": 0,
      "Local Hit Blocks": 0,
      "Local Read Blocks": 0,
      "Local Dirtied Blocks": 0,
      "Local Written Blocks": 0,
      "Temp Read Blocks": 0,
      "Temp Written Blocks": 0
    },
    "Query Identifier": 5963631795316456703,
    "Planning": {
      "Shared Hit Blocks": 0,
      "Shared Read Blocks": 0,
      "Shared Dirtied Blocks": 0,
      "Shared Written Blocks": 0,
      "Local Hit Blocks": 0,
      "Local Read Blocks": 0,
      "Local Dirtied Blocks": 0,
      "Local Written Blocks": 0,
      "Temp Read Blocks": 0,
      "Temp Written Blocks": 0
    },
    "Planning Time": 0.220,
    "Triggers": [
    ],
    "Execution Time": 0.060
  }
```

If I now turn on an RLS defined as
```sql
create policy "Employees can view files for their organisation" on storage.objects for select to authenticated using (
    bucket_id = 'message_files' and
    is_employee_of_organisation(uuid_from_path_level_1(name))
);
```

and execute the same query with an authenticated role, it does an sequential scan of the entire table, leading to an execution time of 16s.

```json
{
    "Plan": {
      "Node Type": "Seq Scan",
      "Parallel Aware": false,
      "Async Capable": false,
      "Relation Name": "objects",
      "Schema": "storage",
      "Alias": "objects",
      "Startup Cost": 0.00,
      "Total Cost": 116988.09,
      "Plan Rows": 235,
      "Plan Width": 16,
      "Actual Startup Time": 6.688,
      "Actual Total Time": 16593.651,
      "Actual Rows": 2,
      "Actual Loops": 1,
      "Output": ["id"],
      "Filter": "((objects.bucket_id = 'message_files'::text) AND is_employee_of_organisation(uuid_from_path_level_1(objects.name)) AND (uuid_from_path_level_2(objects.name) = '7daa4a53-1233-432d-bf11-a9dee8140027'::uuid))",
      "Rows Removed by Filter": 141277,
      "Shared Hit Blocks": 290980,
      "Shared Read Blocks": 0,
      "Shared Dirtied Blocks": 0,
      "Shared Written Blocks": 0,
      "Local Hit Blocks": 0,
      "Local Read Blocks": 0,
      "Local Dirtied Blocks": 0,
      "Local Written Blocks": 0,
      "Temp Read Blocks": 0,
      "Temp Written Blocks": 0
    },
    "Query Identifier": 5963631795316456703,
    "Planning": {
      "Shared Hit Blocks": 7,
      "Shared Read Blocks": 0,
      "Shared Dirtied Blocks": 0,
      "Shared Written Blocks": 0,
      "Local Hit Blocks": 0,
      "Local Read Blocks": 0,
      "Local Dirtied Blocks": 0,
      "Local Written Blocks": 0,
      "Temp Read Blocks": 0,
      "Temp Written Blocks": 0
    },
    "Planning Time": 0.282,
    "Triggers": [
    ],
    "Execution Time": 16601.688
  }
```

Surprisingly, if I add the exact same clause to the query itself and then execute again without rls, the query is planned and execute exactly as I imagined it to be.

```sql
select id
from storage.objects
where bucket_id = 'message_files'
and uuid_from_path_level_2(name) = '7daa4a53-1233-432d-bf11-a9dee8140027'::uuid 
and is_employee_of_organisation(uuid_from_path_level_1(name));
```

```json
{
    "Plan": {
      "Node Type": "Index Scan",
      "Parallel Aware": false,
      "Async Capable": false,
      "Scan Direction": "Forward",
      "Index Name": "ix_storage_obj_entity_id_from_path",
      "Relation Name": "objects",
      "Schema": "storage",
      "Alias": "objects",
      "Startup Cost": 0.42,
      "Total Cost": 723.44,
      "Plan Rows": 235,
      "Plan Width": 16,
      "Actual Startup Time": 0.682,
      "Actual Total Time": 0.798,
      "Actual Rows": 2,
      "Actual Loops": 1,
      "Output": ["id"],
      "Index Cond": "((objects.bucket_id = 'message_files'::text) AND (uuid_from_path_level_2(objects.name) = '7daa4a53-1233-432d-bf11-a9dee8140027'::uuid))",
      "Rows Removed by Index Recheck": 0,
      "Filter": "is_employee_of_organisation(uuid_from_path_level_1(objects.name))",
      "Rows Removed by Filter": 0,
      "Shared Hit Blocks": 9,
      "Shared Read Blocks": 0,
      "Shared Dirtied Blocks": 0,
      "Shared Written Blocks": 0,
      "Local Hit Blocks": 0,
      "Local Read Blocks": 0,
      "Local Dirtied Blocks": 0,
      "Local Written Blocks": 0,
      "Temp Read Blocks": 0,
      "Temp Written Blocks": 0
    },
    "Query Identifier": -275661627852585747,
    "Planning": {
      "Shared Hit Blocks": 7,
      "Shared Read Blocks": 0,
      "Shared Dirtied Blocks": 0,
      "Shared Written Blocks": 0,
      "Local Hit Blocks": 0,
      "Local Read Blocks": 0,
      "Local Dirtied Blocks": 0,
      "Local Written Blocks": 0,
      "Temp Read Blocks": 0,
      "Temp Written Blocks": 0
    },
    "Planning Time": 0.297,
    "Triggers": [
    ],
    "Execution Time": 0.828
  }
```

For me, it seems like the query planner, for some reason, needs to apply the RLS filter with the first part of the query, instead of first reducing the number of rows using an index. For all other tables, the same concept works really well and the planner applies the filter only to the remaining rows. I found a few references in other places where similar behaviour was reported, but the main answers where "its a case to case thing", hence I am asking this in the contexts of supabase. [This SO answer](https://stackoverflow.com/questions/63617827/why-is-row-level-security-rls-not-using-indexes) explains it with a `subquery pullup [that] is happening before RLS policies are taken into account`, but in all honesty I do not know what this means and how to prevent it.

Best
Philipp

**Comment by psteinroe** [comment]
Edit:

If the RLS is set to true, it picks up the index as expected.
```sql
create policy "Employees can view files for their organisation" on storage.objects for select to authenticated using (
   true
);

```
```json
{
    "Plan": {
      "Node Type": "Index Scan",
      "Parallel Aware": false,
      "Async Capable": false,
      "Scan Direction": "Forward",
      "Index Name": "ix_storage_obj_entity_id_from_path",
      "Relation Name": "objects",
      "Schema": "storage",
      "Alias": "objects",
      "Startup Cost": 0.42,
      "Total Cost": 370.92,
      "Plan Rows": 706,
      "Plan Width": 16,
      "Actual Startup Time": 0.023,
      "Actual Total Time": 0.026,
      "Actual Rows": 2,
      "Actual Loops": 1,
      "Output": ["id"],
      "Index Cond": "((objects.bucket_id = 'message_files'::text) AND (uuid_from_path_level_2(objects.name) = '7daa4a53-1233-432d-bf11-a9dee8140027'::uuid))",
      "Rows Removed by Index Recheck": 0,
      "Shared Hit Blocks": 5,
      "Shared Read Blocks": 0,
      "Shared Dirtied Blocks": 0,
      "Shared Written Blocks": 0,
      "Local Hit Blocks": 0,
      "Local Read Blocks": 0,
      "Local Dirtied Blocks": 0,
      "Local Written Blocks": 0,
      "Temp Read Blocks": 0,
      "Temp Written Blocks": 0
    },
    "Query Identifier": 5963631795316456703,
    "Planning": {
      "Shared Hit Blocks": 7,
      "Shared Read Blocks": 0,
      "Shared Dirtied Blocks": 0,
      "Shared Written Blocks": 0,
      "Local Hit Blocks": 0,
      "Local Read Blocks": 0,
      "Local Dirtied Blocks": 0,
      "Local Written Blocks": 0,
      "Temp Read Blocks": 0,
      "Temp Written Blocks": 0
    },
    "Planning Time": 0.281,
    "Triggers": [
    ],
    "Execution Time": 0.050
  }
```

but as soon as I do so much as adding the bucket_id in the RLS, it does a sequential scan again:
```sql
create policy "Employees can view files for their organisation" on storage.objects for select to authenticated using (
   bucket_id = 'message_files'
);
```

```json
{
    "Plan": {
      "Node Type": "Seq Scan",
      "Parallel Aware": false,
      "Async Capable": false,
      "Relation Name": "objects",
      "Schema": "storage",
      "Alias": "objects",
      "Startup Cost": 0.00,
      "Total Cost": 46337.08,
      "Plan Rows": 706,
      "Plan Width": 16,
      "Actual Startup Time": 3.710,
      "Actual Total Time": 7155.941,
      "Actual Rows": 2,
      "Actual Loops": 1,
      "Output": ["id"],
      "Filter": "((objects.bucket_id = 'message_files'::text) AND (uuid_from_path_level_2(objects.name) = '7daa4a53-1233-432d-bf11-a9dee8140027'::uuid))",
      "Rows Removed by Filter": 141295,
      "Shared Hit Blocks": 8878,
      "Shared Read Blocks": 0,
      "Shared Dirtied Blocks": 0,
      "Shared Written Blocks": 0,
      "Local Hit Blocks": 0,
      "Local Read Blocks": 0,
      "Local Dirtied Blocks": 0,
      "Local Written Blocks": 0,
      "Temp Read Blocks": 0,
      "Temp Written Blocks": 0
    },
    "Query Identifier": 5963631795316456703,
    "Planning": {
      "Shared Hit Blocks": 7,
      "Shared Read Blocks": 0,
      "Shared Dirtied Blocks": 0,
      "Shared Written Blocks": 0,
      "Local Hit Blocks": 0,
      "Local Read Blocks": 0,
      "Local Dirtied Blocks": 0,
      "Local Written Blocks": 0,
      "Temp Read Blocks": 0,
      "Temp Written Blocks": 0
    },
    "Planning Time": 0.273,
    "Triggers": [
    ],
    "Execution Time": 7158.675
  }
```

**Comment by psteinroe** [comment]
here is the full list of indices on the storage.objects table:

| indexdef                                                                                                                                                                                                                                                                                                               |
| ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| CREATE UNIQUE INDEX objects_pkey ON storage.objects USING btree (id)                                                                                                                                                                                                                                                   |
| CREATE UNIQUE INDEX bucketid_objname ON storage.objects USING btree (bucket_id, name)                                                                                                                                                                                                                                  |
| CREATE INDEX name_prefix_search ON storage.objects USING btree (name text_pattern_ops)                                                                                                                                                                                                                                 |
| CREATE INDEX ix_storage_obj_uuid_from_filename ON storage.objects USING btree (uuid_from_filename(name)) WHERE (bucket_id = ANY (ARRAY['organisation_favicons'::text, 'profile_images'::text, 'organization_logos'::text, 'contact_widget_images'::text, 'commercial_register_extract'::text, 'identity_card'::text])) |
| CREATE INDEX ix_storage_obj_entity_id_from_path ON storage.objects USING btree (bucket_id, uuid_from_path_level_2(name)) WHERE (bucket_id = ANY (ARRAY['message_files'::text, 'contact_files'::text, 'template_files'::text]))                                                                                         |
| CREATE INDEX ix_storage_obj_message_id_from_path ON storage.objects USING btree (bucket_id, uuid_from_path_level_3(name)) WHERE (bucket_id = 'message_files'::text)                                                                                                                                                    |
| CREATE INDEX ix_storage_obj_org_id_from_path ON storage.objects USING btree (bucket_id, uuid_from_path_level_1(name)) WHERE (bucket_id = ANY (ARRAY['message_files'::text, 'contact_files'::text, 'template_files'::text]))                                                                                            |                                                                                  |

**Comment by psteinroe** [comment]
Also, all of the functions used in either query or indices have proper permissions and volatility set.

**Reply by michelp** [reply]
This is a common problem with RLS.  The planner cannot "look into" the functions you call in policies to determine in advance which rows they will exclude, thus, it must call the functions for every candidate row this typically induces a table scan.  Only functions marked `LEAKPROOF` can be safely called in advance by the planner.

> To specify which rows are visible or modifiable according to a policy, an expression is required that returns a Boolean result. This expression will be evaluated for each row prior to any conditions or functions coming from the user's query. (The only exceptions to this rule are leakproof functions, which are guaranteed to not leak information; the optimizer may choose to apply such functions ahead of the row-security check.)

See: https://www.postgresql.org/docs/current/ddl-rowsecurity.html

It is extremely natural for programmers to want to condense and refactor common policy logic into functions the policy calls, but this defeats the planner's ability to discern whether an index can be checked to exclude candidates before evaluates them against the policy.  This is something core developers are aware of, but PG just isn't there yet on optimizing policies with functions that can leak.

Unfortunately the only good approach is to "invert" your policy, bringing the raw sql logic up into the policy itself and not using the functions.  This can often mean some annoying duplication of code.

**Reply by psteinroe** [reply]
Thanks for the hint! I found out about leakproof a few hours ago and learned that functions can be defined as leakproof too. Setting the uuid_from_path rpcs to be leakproof fixes the issue. From what I understand, the problem is that the rpc is ensuring the text is a valid uuid by testing it against a regexp before casting, which is not a leakproof operation. Since wrapping it into a leakproof function however fixes the issue, I am tempted to leave it like that. Is there anything that could go wrong with just defining the function to be leakproof? 

btw. @thorwebdev I am tempted to write a "common pitfalls with Postgres rls" blog post or documentation piece - is there anyone already working on that within the Supabase team?

**Reply by thorwebdev** [reply]
@psteinroe we're making a push for more content around RLS so this would fit in perfectly! I think that could be great as a guest post on our blog, similar to https://supabase.com/blog/choosing-a-postgres-primary-key .  Let me link you up with @RamiroND
